File: app/src/main/java/za/co/woolworths/financial/services/android/ui/activities/StartupActivity.java
package za.co.woolworths.financial.services.android.ui.activities;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;

import androidx.annotation.VisibleForTesting;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import com.awfs.coordination.R;
import com.google.firebase.analytics.FirebaseAnalytics;

import io.fabric.sdk.android.services.common.CommonUtils;
import za.co.woolworths.financial.services.android.contracts.ConfigResponseListener;
import za.co.woolworths.financial.services.android.contracts.FirebaseManagerAnalyticsProperties;
import za.co.woolworths.financial.services.android.viewmodels.StartupViewModel;
import za.co.woolworths.financial.services.android.viewmodels.StartupViewModelImpl;
import za.co.woolworths.financial.services.android.models.dao.SessionDao;
import za.co.woolworths.financial.services.android.models.dto.ConfigResponse;
import za.co.woolworths.financial.services.android.ui.views.WButton;
import za.co.woolworths.financial.services.android.ui.views.WTextView;
import za.co.woolworths.financial.services.android.ui.views.WVideoView;
import za.co.woolworths.financial.services.android.ui.views.actionsheet.RootedDeviceInfoFragment;
import za.co.woolworths.financial.services.android.util.AuthenticateUtils;
import za.co.woolworths.financial.services.android.util.NetworkManager;
import za.co.woolworths.financial.services.android.util.NotificationUtils;
import za.co.woolworths.financial.services.android.util.Utils;

public class StartupActivity extends AppCompatActivity implements MediaPlayer.OnCompletionListener {

    private StartupViewModel startupViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_startup);
        setSupportActionBar(findViewById(R.id.mToolbar));
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null)
            actionBar.hide();

        startupViewModel = new StartupViewModelImpl(this);
        startupViewModel.setIntent(getIntent());
        Bundle bundle = startupViewModel.getIntent().getExtras();
        if (bundle != null)
            startupViewModel.setPushNotificationUpdate(bundle.getString(NotificationUtils.PUSH_NOTIFICATION_INTENT));

        try {
            startupViewModel.setAppVersion(getPackageManager().getPackageInfo(getPackageName(), 0).versionName);
            startupViewModel.setEnvironment(com.awfs.coordination.BuildConfig.FLAVOR);
        } catch (PackageManager.NameNotFoundException e) {
            startupViewModel.setAppVersion("6.1.0");
            startupViewModel.setEnvironment("QA");
            e.printStackTrace();
        }

        if (startupViewModel.getFirebaseAnalytics() == null)
            startupViewModel.setFirebaseAnalytics(FirebaseAnalytics.getInstance(this));

        if (NetworkManager.getInstance().isConnectedToNetwork(this)) {
            startupViewModel.getFirebaseAnalytics().setUserProperty(StartupViewModelImpl.Companion.getAPP_SERVER_ENVIRONMENT_KEY(), startupViewModel.getEnvironment().isEmpty() ? "prod" : startupViewModel.getEnvironment().toLowerCase());
            startupViewModel.getFirebaseAnalytics().setUserProperty(StartupViewModelImpl.Companion.getAPP_VERSION_KEY(), startupViewModel.getAppVersion());

            setupScreen();
        } else {
            showNonVideoViewWithErrorLayout();
        }

        findViewById(R.id.retry).setOnClickListener(onRetryButtonTapped);

        //Remove old usage of SharedPreferences data.
        Utils.clearSharedPreferences(StartupActivity.this);
        AuthenticateUtils.getInstance(StartupActivity.this).enableBiometricForCurrentSession(true);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        startupViewModel.setIntent(intent);
        initialize();
    }

    private void initialize() {

        startupViewModel.queryServiceGetConfig(new ConfigResponseListener() {
            @Override
            public void onSuccess(ConfigResponse response) {
                startupViewModel.setVideoPlayerShouldPlay(false);

                if (response.configs.enviroment.stsURI == null || response.configs.enviroment.stsURI.isEmpty()) {
                    showNonVideoViewWithErrorLayout();
                    return;
                }

                if (!startupViewModel.isVideoPlaying()) {
                    presentNextScreenOrServerMessage();
                }
            }

            @Override
            public void onFailure(Throwable throwable) {
                showNonVideoViewWithErrorLayout();
            }
        });
    }

    //video player on completion
    @Override
    public void onCompletion(MediaPlayer mp) {

        startupViewModel.setVideoPlaying(false);

        if (!startupViewModel.videoPlayerShouldPlay()) {

            presentNextScreenOrServerMessage();
            mp.stop();

        } else {
            showNonVideoViewWithoutErrorLayout();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        startupViewModel.setAppMinimized(true);
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (Utils.checkForBinarySu() && CommonUtils.isRooted(this) && getSupportFragmentManager() != null) {
            ProgressBar pBar = (ProgressBar) findViewById(R.id.progressBar);

            if (pBar != null)
                pBar.setVisibility(View.GONE);

            Utils.setScreenName(this, FirebaseManagerAnalyticsProperties.ScreenNames.DEVICE_ROOTED_AT_STARTUP);
            RootedDeviceInfoFragment rootedDeviceInfoFragment = RootedDeviceInfoFragment.Companion.newInstance(getString(R.string.rooted_phone_desc));
            rootedDeviceInfoFragment.show(getSupportFragmentManager(), RootedDeviceInfoFragment.class.getSimpleName());
            return;
        }

        if (startupViewModel.isAppMinimized()) {
            startupViewModel.setAppMinimized(false);

            if (startupViewModel.isServerMessageShown()) {
                showNonVideoViewWithoutErrorLayout();
                initialize();
            } else {
                startActivity(new Intent(this, StartupActivity.class));
                finish();
            }
        } else {
            initialize();
        }
    }

    private View.OnClickListener onRetryButtonTapped = new View.OnClickListener() {

        @Override
        public void onClick(View v) {
            if (NetworkManager.getInstance().isConnectedToNetwork(StartupActivity.this)) {
                startupViewModel.getFirebaseAnalytics().setUserProperty(StartupViewModelImpl.Companion.getAPP_SERVER_ENVIRONMENT_KEY(), startupViewModel.getEnvironment().isEmpty() ? "prod" : startupViewModel.getEnvironment().toLowerCase());
                startupViewModel.getFirebaseAnalytics().setUserProperty(StartupViewModelImpl.Companion.getAPP_VERSION_KEY(), startupViewModel.getAppVersion());

                setupScreen();
                initialize();
            } else {
                showNonVideoViewWithErrorLayout();
            }
        }
    };

    private void showVideoView() {
        WVideoView videoView = (WVideoView) findViewById(R.id.activity_wsplash_screen_videoview);
        View noVideoView = (View) findViewById(R.id.splashNoVideoView);
        View serverMessageView = (View) findViewById(R.id.splashServerMessageView);
        RelativeLayout videoViewLayout = (RelativeLayout) findViewById(R.id.videoViewLayout);

        noVideoView.setVisibility(View.GONE);
        serverMessageView.setVisibility(View.GONE);
        videoViewLayout.setVisibility(View.VISIBLE);
        String randomVideo = startupViewModel.getRandomVideoPath();
        Uri videoUri = Uri.parse(randomVideo);

        videoView.setVideoURI(videoUri);
        videoView.start();
        videoView.setOnCompletionListener(this);

        startupViewModel.setVideoPlaying(true);
    }

    private void showNonVideoViewWithErrorLayout() {
        LinearLayout errorLayout = (LinearLayout) findViewById(R.id.errorLayout);
        View noVideoView = (View) findViewById(R.id.splashNoVideoView);
        View serverMessageView = (View) findViewById(R.id.splashServerMessageView);
        RelativeLayout videoViewLayout = (RelativeLayout) findViewById(R.id.videoViewLayout);
        ProgressBar pBar = (ProgressBar) findViewById(R.id.progressBar);

        runOnUiThread(() -> {
            pBar.setVisibility(View.GONE);
            videoViewLayout.setVisibility(View.GONE);
            noVideoView.setVisibility(View.VISIBLE);
            serverMessageView.setVisibility(View.GONE);
            errorLayout.setVisibility(View.VISIBLE);
        });

    }

    private void showNonVideoViewWithoutErrorLayout() {
        LinearLayout errorLayout = (LinearLayout) findViewById(R.id.errorLayout);
        View noVideoView = (View) findViewById(R.id.splashNoVideoView);
        View serverMessageView = (View) findViewById(R.id.splashServerMessageView);
        RelativeLayout videoViewLayout = (RelativeLayout) findViewById(R.id.videoViewLayout);
        ProgressBar pBar = (ProgressBar) findViewById(R.id.progressBar);

        pBar.setVisibility(View.VISIBLE);
        videoViewLayout.setVisibility(View.GONE);
        errorLayout.setVisibility(View.GONE);
        noVideoView.setVisibility(View.VISIBLE);
        serverMessageView.setVisibility(View.GONE);
    }

    private void showServerMessage() {
        LinearLayout errorLayout = (LinearLayout) findViewById(R.id.errorLayout);
        View noVideoView = (View) findViewById(R.id.splashNoVideoView);
        View serverMessageView = (View) findViewById(R.id.splashServerMessageView);
        WTextView serverMessageLabel = (WTextView) findViewById(R.id.messageLabel);
        RelativeLayout videoViewLayout = (RelativeLayout) findViewById(R.id.videoViewLayout);
        ProgressBar pBar = (ProgressBar) findViewById(R.id.progressBar);

        pBar.setVisibility(View.GONE);
        videoViewLayout.setVisibility(View.GONE);
        errorLayout.setVisibility(View.GONE);
        noVideoView.setVisibility(View.GONE);

        serverMessageLabel.setText(startupViewModel.getSplashScreenText());
        WButton proceedButton = findViewById(R.id.proceedButton);
        if (startupViewModel.isSplashScreenPersist()) {
            proceedButton.setVisibility(View.GONE);
        } else {
            proceedButton.setVisibility(View.VISIBLE);
            proceedButton.setOnClickListener(v -> {
                showNonVideoViewWithoutErrorLayout();
                startupViewModel.presentNextScreen();
            });
        }

        serverMessageView.setVisibility(View.VISIBLE);

        startupViewModel.setServerMessageShown(true);
    }

    private boolean isFirstTime() {
        if (Utils.getSessionDaoValue(StartupActivity.this, SessionDao.KEY.SPLASH_VIDEO) == null) {
            return true;
        } else
            return false;
    }

    private void setupScreen() {
        if (isFirstTime()) {
            showVideoView();
        } else {
            showNonVideoViewWithoutErrorLayout();
        }
    }

    private void presentNextScreenOrServerMessage() {
        if (startupViewModel.isSplashScreenDisplay()) {
            showServerMessage();
        } else {
            showNonVideoViewWithoutErrorLayout();
            startupViewModel.presentNextScreen();
        }
    }

    protected void onResume() {
        super.onResume();
        Utils.setScreenName(this, FirebaseManagerAnalyticsProperties.ScreenNames.STARTUP);
        NotificationUtils.clearNotifications(StartupActivity.this);
    }

    @VisibleForTesting
    public boolean testIsFirstTime() {
        return this.isFirstTime();
    }

    @VisibleForTesting
    public String testGetRandomVideos() {
        return startupViewModel.getRandomVideoPath();
    }

//    private void openDeepLinkBackgroundActivity(String isFirstTime) {
//        if (isFirstTime == null || Utils.isAppUpdated(this)) {
//            ScreenManager.presentOnboarding(StartupActivity.this);
//        } else {
//            Intent openBottomActivity = new Intent(this, BottomNavigationActivity.class);
//            openBottomActivity.putExtra(NotificationUtils.PUSH_NOTIFICATION_INTENT, "");
//            startActivity(openBottomActivity);
//            overridePendingTransition(0, 0);
//        }
//    }
}

File: app/src/main/java/za/co/woolworths/financial/services/android/ui/activities/StartupActivity.kt
package za.co.woolworths.financial.services.android.ui.activities

import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.graphics.PorterDuff
import android.media.MediaPlayer
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AppCompatActivity
import com.awfs.coordination.BuildConfig
import com.awfs.coordination.R
import com.google.firebase.analytics.FirebaseAnalytics
import io.fabric.sdk.android.services.common.CommonUtils
import kotlinx.android.synthetic.main.activity_startup.*
import kotlinx.android.synthetic.main.activity_startup_with_message.*
import kotlinx.android.synthetic.main.activity_startup_without_video.*
import za.co.woolworths.financial.services.android.contracts.FirebaseManagerAnalyticsProperties
import za.co.woolworths.financial.services.android.contracts.IResponseListener
import za.co.woolworths.financial.services.android.models.dao.SessionDao
import za.co.woolworths.financial.services.android.models.dto.ConfigResponse
import za.co.woolworths.financial.services.android.ui.views.actionsheet.RootedDeviceInfoFragment
import za.co.woolworths.financial.services.android.ui.views.actionsheet.RootedDeviceInfoFragment.Companion.newInstance
import za.co.woolworths.financial.services.android.util.AuthenticateUtils
import za.co.woolworths.financial.services.android.util.NetworkManager
import za.co.woolworths.financial.services.android.util.NotificationUtils
import za.co.woolworths.financial.services.android.util.Utils
import za.co.woolworths.financial.services.android.viewmodels.StartupViewModel
import za.co.woolworths.financial.services.android.viewmodels.StartupViewModelImpl
import za.co.woolworths.financial.services.android.viewmodels.StartupViewModelImpl.Companion.APP_SERVER_ENVIRONMENT_KEY
import za.co.woolworths.financial.services.android.viewmodels.StartupViewModelImpl.Companion.APP_VERSION_KEY
import java.util.*

class StartupActivity : AppCompatActivity(), MediaPlayer.OnCompletionListener, View.OnClickListener {

    private var startupViewModel: StartupViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_startup)
        setSupportActionBar(mToolbar)
        window?.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        supportActionBar?.hide()

        progressBar?.indeterminateDrawable?.setColorFilter(Color.BLACK, PorterDuff.Mode.MULTIPLY)

        // Disable first time launch splash video screen, remove to enable video on startup
        Utils.sessionDaoSave(SessionDao.KEY.SPLASH_VIDEO, "1")

        startupViewModel = StartupViewModelImpl(this)
        startupViewModel?.apply {
            this.intent = getIntent()
            val bundle = getIntent()?.extras

            pushNotificationUpdate = bundle?.getString(NotificationUtils.PUSH_NOTIFICATION_INTENT)

            try {
                appVersion = packageManager.getPackageInfo(packageName, 0).versionName
                environment = BuildConfig.FLAVOR
            } catch (e: PackageManager.NameNotFoundException) {
                appVersion = "6.1.0"
                environment = "QA"
            }

            firebaseAnalytics = FirebaseAnalytics.getInstance(this@StartupActivity)

            if (NetworkManager.getInstance().isConnectedToNetwork(this@StartupActivity)) {
                firebaseAnalytics?.apply {
                    setUserProperty(APP_SERVER_ENVIRONMENT_KEY, if (environment?.isEmpty() == true) "prod" else environment?.toLowerCase(Locale.getDefault()))
                    setUserProperty(APP_VERSION_KEY, appVersion)
                }
                setupScreen()
            } else {
                showNonVideoViewWithErrorLayout()
            }

            retry?.setOnClickListener(this@StartupActivity)

            //Remove old usage of SharedPreferences data.
            Utils.clearSharedPreferences(this@StartupActivity)
            AuthenticateUtils.getInstance(this@StartupActivity).enableBiometricForCurrentSession(true)
        }
    }

    private fun setupScreen() {
       val isFirstTime: Boolean = isFirstTime()
        if (isFirstTime) {
            showVideoView()
        } else {
            showNonVideoViewWithoutErrorLayout()
        }
    }

    private fun isFirstTime(): Boolean {
        return Utils.getSessionDaoValue(SessionDao.KEY.SPLASH_VIDEO) == null
    }

    private fun showVideoView() {
        splashNoVideoView?.visibility = View.GONE
        splashServerMessageView?.visibility = View.GONE
        videoViewLayout?.visibility = View.VISIBLE

        val randomVideo = startupViewModel?.randomVideoPath
        if (randomVideo?.isNotEmpty() == true) {
            val videoUri = Uri.parse(randomVideo)
            activity_wsplash_screen_videoview?.apply {
                setVideoURI(videoUri)
                start()
                setOnCompletionListener(this@StartupActivity)
            }
            startupViewModel?.isVideoPlaying = false
        }
    }

    private fun showNonVideoViewWithErrorLayout() {
        runOnUiThread {
            progressBar?.visibility = View.GONE
            splashNoVideoView?.visibility = View.GONE
            splashNoVideoView?.visibility = View.VISIBLE
            splashServerMessageView?.visibility = View.GONE
            errorLayout?.visibility = View.VISIBLE
        }
    }

    private fun showNonVideoViewWithoutErrorLayout() {
        progressBar?.visibility = View.VISIBLE
        videoViewLayout?.visibility = View.GONE
        errorLayout?.visibility = View.GONE
        splashNoVideoView?.visibility = View.VISIBLE
        splashServerMessageView?.visibility = View.GONE
    }

    override fun onClick(v: View?) {
        startupViewModel?.apply {
            when (v?.id) {
                R.id.retry -> {
                    if (NetworkManager.getInstance().isConnectedToNetwork(this@StartupActivity)) {
                        firebaseAnalytics?.apply {
                            setUserProperty(APP_SERVER_ENVIRONMENT_KEY, if (environment?.isEmpty() == true) "prod" else environment?.toLowerCase(Locale.getDefault()))
                            setUserProperty(APP_VERSION_KEY, appVersion)
                        }
                        setupScreen()
                        initialize()
                    } else {
                        showNonVideoViewWithErrorLayout()
                    }
                }
            }
        }
    }

    private fun initialize() {
        startupViewModel?.apply {
            queryServiceGetConfig(object : IResponseListener<ConfigResponse?> {
                override fun onSuccess(response: ConfigResponse?) {
                    videoPlayerShouldPlay = false
                    if (response?.configs?.enviroment?.stsURI == null || response.configs.enviroment.stsURI.isEmpty()) {
                        showNonVideoViewWithErrorLayout()
                        return
                    }
                    if (!isVideoPlaying) {
                        presentNextScreenOrServerMessage()
                    }
                }

                override fun onFailure(throwable: Throwable) {
                    showNonVideoViewWithErrorLayout()
                }
            })
        }
    }

    //video player on completion
    override fun onCompletion(mp: MediaPlayer?) {
        startupViewModel?.apply {
            isVideoPlaying = false
            if (!videoPlayerShouldPlay) {
                presentNextScreenOrServerMessage()
                mp?.stop()
            } else {
                showNonVideoViewWithoutErrorLayout()
            }
        }
    }

    private fun presentNextScreenOrServerMessage() {
        if (startupViewModel?.isSplashScreenDisplay == true) {
            showServerMessage()
        } else {
            showNonVideoViewWithoutErrorLayout()
            startupViewModel?.presentNextScreen()
        }
    }

    private fun showServerMessage() {
        progressBar?.visibility = View.GONE
        videoViewLayout?.visibility = View.GONE
        errorLayout?.visibility = View.GONE
        splashNoVideoView?.visibility = View.GONE
        messageLabel?.setText(startupViewModel?.splashScreenText)
        if (startupViewModel?.isSplashScreenPersist == true) {
            proceedButton?.visibility = View.GONE
        } else {
            proceedButton?.visibility = View.VISIBLE
            proceedButton?.setOnClickListener { _: View? ->
                showNonVideoViewWithoutErrorLayout()
                startupViewModel?.presentNextScreen()
            }
        }
        splashServerMessageView?.visibility = View.VISIBLE
        startupViewModel?.isServerMessageShown = true
    }

    override fun onStart() {
        super.onStart()
        if (Utils.checkForBinarySu() && CommonUtils.isRooted(this)) {
            Utils.setScreenName(this, FirebaseManagerAnalyticsProperties.ScreenNames.DEVICE_ROOTED_AT_STARTUP)
            val rootedDeviceInfoFragment = newInstance(getString(R.string.rooted_phone_desc))
            rootedDeviceInfoFragment.show(supportFragmentManager, RootedDeviceInfoFragment::class.java.simpleName)
            return
        }
        startupViewModel?.apply {
            if (isAppMinimized) {
                isAppMinimized = false
                if (isServerMessageShown) {
                    showNonVideoViewWithoutErrorLayout()
                    initialize()
                } else {
                    startActivity(Intent(this@StartupActivity, StartupActivity::class.java))
                    finish()
                }
            } else {
                initialize()
            }
        }
    }

    override fun onStop() {
        super.onStop()
        startupViewModel?.isAppMinimized = true
    }

    override fun onResume() {
        super.onResume()
        Utils.setScreenName(this, FirebaseManagerAnalyticsProperties.ScreenNames.STARTUP)
        NotificationUtils.clearNotifications(this@StartupActivity)
    }

    @VisibleForTesting
    fun testIsFirstTime(): Boolean {
        return isFirstTime()
    }

    @VisibleForTesting
    fun testGetRandomVideos(): String? {
        return startupViewModel?.randomVideoPath
    }
}